### ISSUE TITLE
Parametric polymorphism not expressible in some contexts

### ISSUE THREAD

**Issue by colehaus** [microsoft/pyright#5431]
In the following code snippet, I would expect all variations to type check.

```py
from typing import Callable, Mapping, Protocol, TypeVar

X = TypeVar("X")

# Fine in pyright and mypy
def id(a: X) -> X:
  return a
  
class Id(Protocol[X]):
    def __call__(self, x: X) -> X: ...

# Error in both pyright and mypy
y: Id[X] = lambda a: a
# Error in pyright, fine in mypy
z: Callable[[X],X] = lambda a: a
```

However, only `id` type checks in `pyright 1.1.316`. The `y` and `z` lines both cause pyright to complain that 'Type variable "X" has no meaning in this context'. `mypy 1.4.1` is able to type check the `z` line correctly. (I'm most interested in the `Callable` case, but I thought it was worth checking to see if the `Protocol` approach worked for some reason.)

**Comment by erictraut** [comment]
Pyright (and mypy) are both correct to generate an error in this case. You're using a type variable in a context where it has no meaning. Type variables must be scoped to a generic function, generic class, or generic type alias. You're using them outside of these contexts. For more details about type variable scoping, refer to PEP 484.

**Comment by colehaus** [comment]
So mypy is "wrong" (i.e. too permissive) in accepting `z`? 

The part of PEP 484 you're referring to is https://peps.python.org/pep-0484/#scoping-rules-for-type-variables ? I see it mentions some special casing of type variables for functions and classes but nothing that explicitly pertains to this behavior, as far as I can tell.

I think part of my confusion here is the "has no meaning" language which does not seem accurate to me? `z: Callable[[X],X] = lambda a: a` seems to me to have a fairly intelligible meaning—this function takes a value of type X and returns a value of type X. Any type-changing function would be forbidden by the declaration (i.e. there's a coherent, useful meaning because this type permits some values and forbids others). In Haskell (I know the Python type system is not trying to be Haskell, but Haskell's type system is a sound one that I'm familiar with), you can express types like this:

```hs
id :: forall a. a -> a
id x = x

ids :: forall a. [a -> a]
ids = [id, \x -> x]
```
and they're considered meaningful.

Thanks!

**Comment by erictraut** [comment]
You are not alone in being confused about the scoping rules for type variables in the Python type system. While PEP 484 made an attempt to define the TypeVar scoping rules, it left some aspects underspecified. We tried to clear up some of this confusion in PEP 695 (which I authored).
